// Copyright (c) YugaByte, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations
// under the License.
//

syntax = "proto3";

package yb.tserver;

option java_package = "org.yb.tserver";

import "yb/common/common.proto";
import "yb/common/wire_protocol.proto";
import "yb/master/master.proto";

service PgClientService {
  rpc GetDatabaseInfo(PgGetDatabaseInfoRequestPB) returns (PgGetDatabaseInfoResponsePB);
  rpc IsInitDbDone(PgIsInitDbDoneRequestPB) returns (PgIsInitDbDoneResponsePB);
  rpc OpenTable(PgOpenTableRequestPB) returns (PgOpenTableResponsePB);
  rpc ReserveOids(PgReserveOidsRequestPB) returns (PgReserveOidsResponsePB);
}

message PgGetDatabaseInfoRequestPB {
  uint32 oid = 1;
}

message PgGetDatabaseInfoResponsePB {
  AppStatusPB status = 1;

  master.GetNamespaceInfoResponsePB info = 2;
}

message PgIsInitDbDoneRequestPB {
}

message PgIsInitDbDoneResponsePB {
  AppStatusPB status = 1;
  bool done = 2;
}

message PgOpenTableRequestPB {
  string table_id = 1;
}

message PgTablePartitionsPB {
  uint32 version = 1;
  repeated bytes keys = 2;
}

message PgOpenTableResponsePB {
  AppStatusPB status = 1;

  master.GetTableSchemaResponsePB info = 2;
  PgTablePartitionsPB partitions = 3;
}

message PgReserveOidsRequestPB {
  uint32 database_oid = 1;
  uint32 next_oid = 2;
  uint32 count = 3;
}

message PgReserveOidsResponsePB {
  AppStatusPB status = 1;

  uint32 begin_oid = 2;
  uint32 end_oid = 3;
}
